infra/Pos/Slotting/Util.hs

```diff
@@ -42,57 +42,57 @@ import           Pos.Shutdown           (HasShutdownContext, runIfNotShutdown)
+    -- Maybe epoch slotting data.
+    pure $ do
+      epochSlottingData <- mEpochSlottingData
+      pure $ getSlotStartPureFromEpoch si epochSlottingData
 
 -- | Pure timestamp calculation for a given slot.
--- If `imprecise` is true, then we assume that the slot duration doesn't change.
--- That allows us to compute a timestamp for slots earlier the penultimate epoch.
-getSlotStartPure :: Timestamp -> Bool -> SlotId -> SlottingData -> Maybe Timestamp
-getSlotStartPure systemStart imprecise SlotId{..} SlottingData{..} = do
-    if | imprecise && siEpoch < sdPenultEpoch ->
-         Just $ slotTimestamp siSlot $ extrapolateSlottingData siEpoch
-       | siEpoch == sdPenultEpoch -> Just $ slotTimestamp siSlot sdPenult
-       | siEpoch == sdPenultEpoch + 1 -> Just $ slotTimestamp siSlot sdLast
-       | otherwise -> Nothing
-  where
-    -- Extrapolate slotting data for arbitrary epochs
-    extrapolateSlottingData desiredEpoch =
-      let
-        -- Assuming these durations stay constant
-        epochDuration = esdStartDiff sdLast `subTimeDiffSafe` esdStartDiff sdPenult
-        slotDuration = esdSlotDuration sdPenult
-        msDiff = fromIntegral (toInteger epochDuration * (toInteger desiredEpoch - toInteger sdPenultEpoch))
-      in
-        EpochSlottingData { esdSlotDuration = slotDuration
-                          , esdStartDiff = msDiff + (esdStartDiff sdPenult)
-                          }
-    -- Calculate timestamp normally
-    slotTimestamp (getSlotIndex -> locSlot) EpochSlottingData{..} =
-        esdStartDiff `addTimeDiffToTimestamp`
-        (systemStart + Timestamp (fromIntegral locSlot * convertUnit esdSlotDuration))
+getSlotStartPureFromEpoch :: SlotId -> EpochSlottingData -> Timestamp
+getSlotStartPureFromEpoch = computeSlotStart
+
+-- | Pure timestamp calculation for a given slot.
+getSlotStartPure :: SlotId -> SlottingData -> Maybe Timestamp
+getSlotStartPure si sd = getSlotStartPureFromEpoch si <$> esd
+  where esd = lookupEpochSlottingData (siEpoch si) sd
 
 -- | Get timestamp when given slot starts empatically, which means
 -- that function throws exception when slot start is unknown.
 getSlotStartEmpatically
     :: (MonadSlotsData m, MonadThrow m)
-    => SlotId -> m Timestamp
+    => SlotId
+    -> m Timestamp
 getSlotStartEmpatically slot =
     getSlotStart slot >>= maybeThrow (SEUnknownSlotStart slot)
 
+-- | Get epoch slotting data empatically, which means
+-- that function throws exception when epoch index is unknown.
+getEpochSlottingDataEmpatically
+    :: (MonadSlotsData m, MonadThrow m)
+    => EpochIndex
+    -> m EpochSlottingData
+getEpochSlottingDataEmpatically epochIndex =
+    getEpochSlottingDataM epochIndex >>= maybeThrow (SEUnknownEpoch epochIndex)
+
 -- | Get last known slot duration.
-getLastKnownSlotDuration :: MonadSlotsData m => m Millisecond
-getLastKnownSlotDuration = esdSlotDuration . sdLast <$> getSlottingData
+getLastKnownSlotDuration
```

Comment:
I think with current renamings it's better to rename it to `getNextEpochSlotDuration` (and update the comment above).